### PR TITLE
NegativeArraySizeException not thrown from StringBuffer constructor

### PR DESCRIPTION
**Pull request by babsingh**
NegativeArraySizeException is not thrown from the StringBuffer
constructor that takes capacity as an argument. In the constructor, the
capacity argument is modified before the byte/char array is created. If
the capacity argument is Integer.MIN_VALUE (-2147483648), then capacity
*= 2 will yield a non-negative number due to overflow. Thus, we will
fail to throw NegativeArraySizeException. A new check has been added to
assure that NegativeArraySizeException is thrown if capacity argument is
less than 0.

The implementation of similar constructor in StringBuilder has been
updated to maintain consistency between StringBuffer and StringBuilder.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

File: jcl/src/java.base/share/classes/java/lang/StringBuffer.java [hunk 1]
```diff
@@ -156,36 +156,39 @@ public StringBuffer() {
...
+	 * will yield a non-negative number due to overflow. We will fail to throw
+	 * NegativeArraySizeException. The check below will assure that
+	 * NegativeArraySizeException is thrown if capacity argument is less than 0.
+	 */
+	if (capacity < 0) {
+		throw new NegativeArraySizeException();
+	}
+	
+	/*[IF !Sidecar19-SE]*/
+	this.capacity = capacity;
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/StringBuffer.java]: This should be at the end of the method as we want the allocation of the array to occur before the capacity is assigned.

Comment (babsingh) [reply jcl/src/java.base/share/classes/java/lang/StringBuffer.java]: fixed.

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux

**Comment by babsingh** [comment]
`public StringBuilder(int capacity) { ` and `public StringBuffer(int capacity) {` were identical before except the `NegativeArraySizeException` error-check.

I feel that the old constructor implementation is inefficient: more bytecodes + redundant code.

Old:
```
public StringBuffer(int capacity) {
	if (capacity < 0) {
		throw new NegativeArraySizeException();
	}

        if (String.enableCompression) {
                if (capacity <= MAX_CAPACITY) {
                        value = new byte[capacity];
                } else {
                        // K05df = Unable to allocate an array of the specified capacity. The maximum supported capacity is Integer.MAX_VALUE / 2.
                        throw new OutOfMemoryError(com.ibm.oti.util.Msg.getString("K05df")); //$NON-NLS-1$
                }
        } else {
                if (capacity <= MAX_CAPACITY) {
                        value = new byte[capacity * 2];
                } else {
                        // K05df = Unable to allocate an array of the specified capacity. The maximum supported capacity is Integer.MAX_VALUE / 2.
                        throw new OutOfMemoryError(com.ibm.oti.util.Msg.getString("K05df")); //$NON-NLS-1$
                }
        }

}
```

New:
```
public StringBuffer(int capacity) {
	if (capacity < 0) {
		throw new NegativeArraySizeException();
	}
	int arraySize = capacity;
	
	if (capacity <= MAX_CAPACITY) {
		if (!String.enableCompression) {
			arraySize = capacity * 2;
		}
		value = new byte[arraySize];
	} else {
		/*[MSG "K05df", "Unable to allocate an array of the specified capacity. The maximum supported capacity is Integer.MAX_VALUE / 2."]*/
		throw new OutOfMemoryError(com.ibm.oti.util.Msg.getString("K05df")); //$NON-NLS-1$
	}
}
```

I agree with reusing the implementation if a common implementation is used from `AbstractStringBuilder`.

**Comment by fjeremic** [comment]
> I feel that the old constructor implementation is inefficient: more bytecodes + redundant code.

I suppose my bias here is thinking from the JIT perspective as `String.enableCompression` is a `static final boolean` and will be folded away when the method gets JIT compiled. So really the `if (String.enableCompression)` block or the `else` block will get completely folded away. So this is a moot point.

In either case, the point still stands. I'm find with whatever implementation as long as we keep the two classes consistent. If the new implementation you described above is better from the interpreter point of view we should also make the same change on StringBuilder and keep the two constructions identical. #770 is a side issue which can avoid these problems in the future. However that needs to be discussed outside the scope of this PR. Hope I made sense.

**Comment by andrewcraik** [comment]
The argument about more bytecodes being executed in the old implementation I don't think holds - there are two comparisons in both implementations. The benefit of the old implementation is that the case for compressed strings is separate from non-compressed strings. Further the direct use of the field does not require a temporary value to be held, propagated and eliminated by the JIT. In this case the difference would only show up at low opt levels (eg cold). The redundant code is the duplicate exception throw. In general we need to be very careful in these classes cleaning up the code in ways that could make it less performant. In this case someone will need to the check the JIT compilation of this method at cold and warm to see if the extra temp reduces the performance of the code compared to the direct field read in the old implementation which needs less work from the optimizer.

If the optimizer handles the new code as well I am fine with the clean-up.

**Comment by DanHeidinga** [comment]
Adding @andrewcraik as a reviewer so he can sign off on the validated jit logs for

>In this case someone will need to the check the JIT compilation of this method at cold and warm to see if the extra temp reduces the performance of the code compared to the direct field read in the old implementation which needs less work from the optimizer.

**Comment by babsingh** [comment]
@fjeremic The implementation of similar constructor in `StringBuilder` has been
updated to maintain consistency between `StringBuffer` and `StringBuilder`.

**Comment by babsingh** [comment]
Requested @jdmpapin help with collecting compilation logs with and without changes.

**Comment by fjeremic** [comment]
> @fjeremic The implementation of similar constructor in StringBuilder has been
> updated to maintain consistency between StringBuffer and StringBuilder.

Awesome! That was very fast. Thanks for addressing these concerns.

**Comment by DanHeidinga** [comment]
@babsingh /  @jdmpapin - have the compilation logs been collected for this change?

**Comment by babsingh** [comment]
@DanHeidinga compilation logs haven't been collected. awaiting for benchmark from @jdmpapin.

these changes may not be accepted if it's unfavorable to JIT's performance. shall i create a new pull request with minimal changes to just add the error check?

**Comment by babsingh** [comment]
Devin's (@jdmpapin) analysis:

I ran these with reduced "fanout" to avoid logging 12+ compilations, but for StringBuilder I only reduced it to 4 so there are more compilations in those logs. **The performance was roughly on par.**
The new arraySize variable is reliably eliminated. There are still stores but no loads at the time of GRA, and those stores are cleaned up by the immediately subsequent passes of global dead store elimination and dead trees elimination. The early compilation for StringBuffer has two remaining occurrences of the variable, but both in a cold loop. That compilation also attempted to stack allocate the StringBuffer, which it shouldn't do, but that was because the escape point had frequency just below 2500, while it was just above 2500 in the corresponding compilation of the original code.

Compilation logs for `StringBuilder`:
1. [builder.old.log.zip](https://github.com/eclipse/openj9/files/1568433/builder.old.log.zip)
2. [builder.new.log.zip](https://github.com/eclipse/openj9/files/1568434/builder.new.log.zip)

Compilation logs for `StringBuffer`:
1. [buffer.old.log.zip](https://github.com/eclipse/openj9/files/1568435/buffer.old.log.zip)
2. [buffer.new.log.zip](https://github.com/eclipse/openj9/files/1568436/buffer.new.log.zip)

**Comment by DanHeidinga** [comment]
Thanks @jdmpapin for the analysis!

@andrewcraik Is that sufficient for you to sign off on this PR?

**Comment by andrewcraik** [comment]
changes are sufficient - change looks good.

**Comment by DanHeidinga** [comment]
Jenkins test sanity plinux

**Comment by fjeremic** [comment]
@babsingh I think you will need to rebase your branch on the latest master as we had a breaking change merged recently related to `OMRRecognizedCallTransformer`. This is why the PPCLE build is failing. A rebase will resolve this issue and then we can re-launch the Jenkins testing.

**Comment by babsingh** [comment]
Rebased.

**Comment by fjeremic** [comment]
Jenkins test sanity plinux

**Pull request merged**